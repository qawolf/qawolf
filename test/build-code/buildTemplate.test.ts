import {
  buildScriptTemplate,
  buildTestTemplate,
  buildValidVariableName,
} from '../../src/build-code/buildTemplate';

describe('buildTemplate', () => {
  it('builds script template', () => {
    const template = buildScriptTemplate({
      name: 'myScript',
      url: 'www.qawolf.com',
    });
    expect(template).toMatchSnapshot();
  });

  it('builds test template', () => {
    const template = buildTestTemplate({
      name: 'myTest',
      url: 'www.qawolf.com',
    });
    expect(template).toMatchSnapshot();
  });

  it('throws an error if device does not exist', () => {
    const testFn = (): string =>
      buildTestTemplate({
        device: 'unknown',
        name: 'myTest',
        url: 'www.qawolf.com',
      });
    expect(testFn).toThrowError('Device unknown not available in Playwright');
  });

  it('includes device if specified', () => {
    const template = buildScriptTemplate({
      device: 'iPhone 11',
      name: 'myScript',
      url: 'www.qawolf.com',
    });
    expect(template).toMatchSnapshot();

    const template2 = buildTestTemplate({
      device: 'iPad Mini',
      name: 'myTest',
      url: 'www.qawolf.com',
    });
    expect(template2).toMatchSnapshot();
  });

  it('sets state if statePath specified', () => {
    const template = buildScriptTemplate({
      name: 'myScript',
      statePath: 'admin.json',
      url: 'www.qawolf.com',
    });
    expect(template).toMatchSnapshot();

    const template2 = buildTestTemplate({
      name: 'myTest',
      statePath: 'admin.json',
      url: 'www.qawolf.com',
    });
    expect(template2).toMatchSnapshot();
  });

  it('corrects invalid script names if posssible', () => {
    const template = buildScriptTemplate({
      name: 'my-script',
      url: 'www.qawolf.com',
    });
    expect(template).toMatchSnapshot();
  });

  it('throws errors for script names that will never be valid', () => {
    expect(() =>
      buildScriptTemplate({
        name: 'break',
        url: 'www.qawolf.com',
      }),
    ).toThrowError();
  });
});

describe('buildValidVariableName', () => {
  it('returns valid variable names as is', () => {
    expect(buildValidVariableName('myScript')).toBe('myScript');
    expect(buildValidVariableName('my_script')).toBe('my_script');
    expect(buildValidVariableName('myScript1')).toBe('myScript1');
    expect(buildValidVariableName('$myScript')).toBe('$myScript');
  });

  it('returns corrected invalid variable names', () => {
    expect(buildValidVariableName('my-script')).toBe('myScript');
    expect(buildValidVariableName('my script')).toBe('myScript');
  });

  it('throws error if variable name cannot be corrected', () => {
    expect(() => buildValidVariableName('return')).toThrowError('return');
    expect(() => buildValidVariableName('1myScript')).toThrowError('1myScript');
  });
});
