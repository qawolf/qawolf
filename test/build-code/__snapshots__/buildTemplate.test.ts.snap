// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTemplate builds script template 1`] = `
"const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/myScript.json\\");

const myScript = async context => {
  let page = await context.newPage();
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
};

exports.myScript = myScript;

if (require.main === module) {
  (async () => {
    const browser = await qawolf.launch();
    const context = await browser.newContext();
    await qawolf.register(context);
    await myScript(context);
    await qawolf.stopVideos();
    await browser.close();
  })();
}"
`;

exports[`buildTemplate builds test template 1`] = `
"const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/myTest.json\\");

let browser;
let page;

beforeAll(async () => {
  browser = await qawolf.launch();
  const context = await browser.newContext();
  await qawolf.register(context);
  page = await context.newPage();
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test('myTest', async () => {
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
});"
`;

exports[`buildTemplate corrects invalid script names if posssible 1`] = `
"const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/my-script.json\\");

const myScript = async context => {
  let page = await context.newPage();
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
};

exports.myScript = myScript;

if (require.main === module) {
  (async () => {
    const browser = await qawolf.launch();
    const context = await browser.newContext();
    await qawolf.register(context);
    await myScript(context);
    await qawolf.stopVideos();
    await browser.close();
  })();
}"
`;

exports[`buildTemplate includes device if specified 1`] = `
"const { devices } = require(\\"playwright\\");
const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/myScript.json\\");
const device = devices[\\"iPhone 11\\"];

const myScript = async context => {
  let page = await context.newPage();
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
};

exports.myScript = myScript;

if (require.main === module) {
  (async () => {
    const browser = await qawolf.launch();
    const context = await browser.newContext({
      userAgent: device.userAgent,
      viewport: device.viewport
    });
    await qawolf.register(context);
    await myScript(context);
    await qawolf.stopVideos();
    await browser.close();
  })();
}"
`;

exports[`buildTemplate includes device if specified 2`] = `
"const { devices } = require(\\"playwright\\");
const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/myTest.json\\");
const device = devices[\\"iPad Mini\\"];

let browser;
let page;

beforeAll(async () => {
  browser = await qawolf.launch();
  const context = await browser.newContext({
      userAgent: device.userAgent,
      viewport: device.viewport
    });
  await qawolf.register(context);
  page = await context.newPage();
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test('myTest', async () => {
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
});"
`;

exports[`buildTemplate sets state if statePath specified 1`] = `
"const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/myScript.json\\");

const myScript = async context => {
  let page = await context.newPage();
  await qawolf.setState(page, \\"admin.json\\");
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
};

exports.myScript = myScript;

if (require.main === module) {
  (async () => {
    const browser = await qawolf.launch();
    const context = await browser.newContext();
    await qawolf.register(context);
    await myScript(context);
    await qawolf.stopVideos();
    await browser.close();
  })();
}"
`;

exports[`buildTemplate sets state if statePath specified 2`] = `
"const qawolf = require(\\"qawolf\\");
const selectors = require(\\"../selectors/myTest.json\\");

let browser;
let page;

beforeAll(async () => {
  browser = await qawolf.launch();
  const context = await browser.newContext();
  await qawolf.register(context);
  page = await context.newPage();
});

afterAll(async () => {
  await qawolf.stopVideos();
  await browser.close();
});

test('myTest', async () => {
  await qawolf.setState(page, \\"admin.json\\");
  await page.goto(\\"www.qawolf.com\\");
  await qawolf.create();
});"
`;
